draft-ietf-httpbis-connect-tcp.md

```diff
@@ -62,41 +64,47 @@ This specification describes an alternative mechanism for proxying TCP in HTTP.
 
 # Specification
 
-A template-driven TCP transport proxy for HTTP is identified by a URI Template {{!RFC6570}} containing variables named "target_host" and "target_port".  This URI Template and its variable values MUST meet all the same requirements as for UDP proxying ({{!RFC9298, Section 2}}), and are subject to the same validation rules.  The client MUST substitute the destination host and port number into this template to produce the request URI.
+A template-driven TCP transport proxy for HTTP is identified by a URI Template {{!RFC6570}} containing variables named "target_host" and "target_port".  This URI Template and its variable values MUST meet all the same requirements as for UDP proxying ({{!RFC9298, Section 2}}), and are subject to the same validation rules.  The client MUST substitute the destination host and port number into this template to produce the request URI.  The derived URI serves as the destination of a Capsule Protocol connection using the Upgrade Token "connect-tcp" (see registration in {#new-upgrade-token}).
+
+When using "connect-tcp", TCP payload data is sent in the payload of a new Capsule Type named DATA (see registration in {{data-capsule}}).  The ordered concatenation of DATA capsule payloads represents the TCP payload data.
+
+Intermediaries MAY split or merge DATA capsules.  Future CPUTs may use DATA capsules to represent their main data stream, if applicable.
```

Comment:
Thinking about this... should we add a `FLUSH` capsule to force intermediates to flush data and prevent proxies from buffering / concatenating the payload?

Reply:
I agree, we probably don't want intermediaries reordering DATA chunks around other capsules in the stream.  I've added the word "adjacent" to prevent that.

This limitation makes me question the value of the split/merge permission.  I think it makes sense for consistency with H2 DATA frames, but I don't know why anyone would actually implement it.

In addition to the TCP payload data stream, TCP also has a bunch of other obscure features like the "Push" flag (which is pretty much like a flush) and "Urgent" (which is even weirder).  We mostly ignore that stuff, and Classic HTTP CONNECT just bleaches it all.

With Capsules we have the option of bringing some of that stuff back, but I think we can leave that all to future extensions.

Reply:
AFAIK, the push flag is heavily used and added to TCP flags of the packet containing the last bit of data sent from the userland, so it would be unfortunate to lose it.

Notably, I want to avoid situation where things work without when communicating directly, but stop working (or work in degraded fashion) with added intermediary, because we start dropping flags for no reason.

Reply:
I'm pretty sure that the "push" flag is already "bleached" by all existing implementations of HTTP CONNECT over TLS, so I don't think we need to reintroduce it in this draft.  I believe we're in a good position to add support for "push" later if needed.

Reply:
Sure, fine with me. This was mostly a side-comment wrt merged DATA capsules, anyway.